src/coreComponents/constitutive/fluid/DeadOilFluid.hpp

```diff
@@ -0,0 +1,714 @@
+  if( ipWater >= 0 )
+  {
+    // note: double check, but I think std::exp is allowed in kernel
+    real64 const expCompDeltaPres = std::exp( -m_waterCompressibility * ( pressure - m_waterRefPressure ) );
+    real64 const dExpCompDeltaPres_dPres = -m_waterCompressibility * expCompDeltaPres;
+    real64 const denom = m_waterFormationVolFactor * expCompDeltaPres;
+    real64 const dDenom_dPres = m_waterFormationVolFactor * dExpCompDeltaPres_dPres;
+    real64 const denomInv = 1.0 / denom;
+    phaseMassDens[ipWater] = m_surfacePhaseMassDensity[ipWater] * denomInv;
+    dPhaseMassDens_dPres[ipWater] = -dDenom_dPres * phaseMassDens[ipWater] * denomInv;
+  }
+}
+
+GEOSX_HOST_DEVICE
+GEOSX_FORCE_INLINE
+void DeadOilFluidUpdate::computeViscosities( real64 pressure,
+                                             arraySlice1d< real64, 0 > const & phaseVisc ) const
+{
+  // this derivative will be discarded
+  real64 derivative = 0.0;
+
+  // 1. Hydrocarbon phases: look up in the viscosity tables
+
+  for( localIndex iph = 0; iph < m_hydrocarbonPhaseOrder.size(); ++iph )
+  {
+    // get the phase index
+    localIndex const ip = m_hydrocarbonPhaseOrder[iph];
+    // interpolate in the table to get the phase viscosity, and discard the derivative
+    m_viscosityTables[iph].compute( &pressure, phaseVisc[ip], &derivative );
+  }
+
+  // 2. Water phase: use the constant viscosity provided by the user
+
+  using PT = DeadOilFluid::PhaseType;
+  integer const ipWater = m_phaseOrder[PT::WATER];
+
+  // if water is present
+  if( ipWater >= 0 )
+  {
+    // just assign the viscosity value
+    phaseVisc[ipWater] = m_waterViscosity;
+  }
+}
+
+GEOSX_HOST_DEVICE
+GEOSX_FORCE_INLINE
+void DeadOilFluidUpdate::computeViscosities( real64 pressure,
+                                             arraySlice1d< real64 > const & phaseVisc,
+                                             arraySlice1d< real64 > const & dPhaseVisc_dPres,
+                                             arraySlice2d< real64 > const & dPhaseVisc_dGlobalCompFrac ) const
+{
+  for( double & val : dPhaseVisc_dGlobalCompFrac )
+  {
+    val = 0.0;
+  }
+
+  // 1. Hydrocarbon phases: look up in the viscosity tables
+
+  for( localIndex iph = 0; iph < m_hydrocarbonPhaseOrder.size(); ++iph )
+  {
+    // get the phase index
+    localIndex const ip = m_hydrocarbonPhaseOrder[iph];
+    // interpolate in the table to get the phase viscosity and derivatives
+    m_viscosityTables[iph].compute( &pressure, phaseVisc[ip], &(dPhaseVisc_dPres)[ip] );
+  }
+
+  // 2. Water phase: use the constant viscosity provided by the user
+
+  using PT = DeadOilFluid::PhaseType;
+  integer const ipWater = m_phaseOrder[PT::WATER];
+
+  // if water is present
+  if( ipWater >= 0 )
+  {
+    // just assign the viscosity value
+    phaseVisc[ipWater] = m_waterViscosity;
+    dPhaseVisc_dPres[ipWater] = 0.0;
+  }
+}
+
+GEOSX_HOST_DEVICE
+GEOSX_FORCE_INLINE
+void DeadOilFluidUpdate::compute( real64 pressure,
```

Comment:
So you compute point by point? The loop is higher, isn't it?

Reply:
Yes this function computes the update point by point. The RAJA loop that calls `update` which itself calls `compute` is here: https://github.com/GEOSX/GEOSX/blob/5c555f36e7ce939f086e5884306c031acfb77ebb/src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBaseKernels.hpp#L127-L144